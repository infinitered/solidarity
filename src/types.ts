import { GluegunRunContext } from 'gluegun'
export const solidarity = {
  binaryExists: require('./extensions/functions/binaryExists'),
  getSolidaritySettings: require('./extensions/functions/getSolidaritySettings'),
  setSolidaritySettings: require('./extensions/functions/setSolidaritySettings'),
  checkRequirement: require('./extensions/functions/checkRequirement'),
  updateRequirement: require('./extensions/functions/updateRequirement'),
  skipRule: require('./extensions/functions/skipRule'),
  checkENV: require('./extensions/functions/checkENV'),
  checkCLI: require('./extensions/functions/checkCLI'),
  checkDir: require('./extensions/functions/checkDir'),
  checkFile: require('./extensions/functions/checkFile'),
  updateVersions: require('./extensions/functions/updateVersions'),
  removeNonVersionCharacters: require('./extensions/functions/removeNonVersionCharacters'),
  getVersion: require('./extensions/functions/getVersion'),
  getLineWithVersion: require('./extensions/functions/getLineWithVersion')
}

export interface SolidarityPlugin {
  name: string
  description: string
  snapshot: string | SnapshotType
}

export interface SolidarityRunContext extends GluegunRunContext {
  solidarity: typeof solidarity
  _pluginsList: Array<SolidarityPlugin & {templateDirectory: string}>
  addPlugin: (config: SolidarityPlugin) => void
  printSeparator: () => void
  outputMode: SolidarityOutputMode
}

export type SnapshotType = (context: SolidarityRunContext) => Promise<void>

export interface SolidarityRule {
  rule: 'cli' | 'env' | 'dir' | 'file'
  binary?: string
  semver?: string
  version?: string
  line?: string | number
  variable?: string
  location?: string
  error?: string
  platform?: string | string[]
  matchIndex?: number
}

export type SolidarityRequirement = SolidarityRule[]

export interface SolidaritySettings {
  requirements: object
  config: object
}

export enum SolidarityOutputMode {
  MODERATE,
  VERBOSE,
  SILENT
}
