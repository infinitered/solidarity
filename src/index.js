const { build } = require('gluegun')
const {dependenciesMet} = require('gluegun/sniff')
if(dependenciesMet==1){
    module.exports = async () => {
      // setup the runtime
      build()
        .brand('solidarity')
        .src(__dirname)
        // It might also be worth checking global
        // `npm config get prefix` returns '/usr/local'
        // global npm is stored /usr/local/lib/node_modules
        .plugins('./node_modules', { matching: 'solidarity-*', hidden: true })
        // for testing - force load a plugin
        // .plugin('../solidarity-react-native')
        .create()
        .run()
    }
}
