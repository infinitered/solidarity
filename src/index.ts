import { build } from 'gluegun'

module.exports = async (): Promise<void> => {
  // setup the runtime
  build()
    .brand('solidarity')
    .src(__dirname)
    // local installs
    .plugins('./node_modules', { matching: 'solidarity-*', hidden: true })
    // global installs
    .plugins('/usr/local/lib/node_modules', { matching: 'solidarity-*', hidden: true })
    // for testing - force load a local plugin
    // .plugin('../solidarity-react-native')
    .create()
    .run()
}

export * from './types'
