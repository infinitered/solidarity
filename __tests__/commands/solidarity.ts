import checkCommand from '../../src/commands/solidarity'
import { SolidarityOutputMode } from '../../src/types'

const noConfigSolidarity = {
  checkRequirement: jest.fn(),
  getSolidaritySettings: jest.fn(() => ({}))
}

const verboseConfigSolidarity = {
  ...noConfigSolidarity,
  getSolidaritySettings: jest.fn(() => ({
    config: {
      output: 'verbose'
    }
  }))
}

const configNoOutput = {
  ...noConfigSolidarity,
  getSolidaritySettings: jest.fn(() => ({
    config: {}
  }))
}

const mockContext = {
  outputMode: null,
  print: {
    error: jest.fn(),
    success: jest.fn()
  },
  parameters: {
    options: {}
  },
  solidarity: noConfigSolidarity
}

const verboseMockContext = {
  ...mockContext,
  solidarity: verboseConfigSolidarity
}

const noConfigMockContext = {
  ...mockContext,
  solidarity: configNoOutput
}

test('Snapshot check default command', () => {
  expect(checkCommand).toMatchSnapshot()
})

test('Enforce required properties', () => {
  expect(checkCommand.description).toBeTruthy()
  expect(checkCommand.run).toBeTruthy()
  expect(typeof checkCommand.run).toBe('function')
})

test('check base solidarity run', async () => {
  const result = await checkCommand.run(mockContext)
  expect(result).toMatchSnapshot()
  expect(mockContext.outputMode).toBe(SolidarityOutputMode.MODERATE)
  expect(mockContext.solidarity.getSolidaritySettings.mock.calls.length).toBe(1)
  expect(mockContext.print.success.mock.calls.length).toBe(2)
})

test('check solidarity verbose run', async () => {
  const result = await checkCommand.run(verboseMockContext)
  expect(verboseMockContext.outputMode).toBe(SolidarityOutputMode.VERBOSE)
})

test('check solidarity config no output run', async () => {
  const result = await checkCommand.run(noConfigMockContext)
  expect(noConfigMockContext.outputMode).toBe(SolidarityOutputMode.MODERATE)
})
