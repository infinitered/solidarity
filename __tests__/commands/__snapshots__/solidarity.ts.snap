// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Snapshot check default command 1`] = `
Object {
  "description": "Check environment against solidarity rules",
  "run": [Function],
}
`;
